### DISCUSSION TITLE
How to exclusively show all private members

### DISCUSSION THREAD

**Discussion by ItsDrike** [mkdocstrings/mkdocstrings#699] [Q&A] [answered]
I want to split my documentation into 2 sections, where the first one represents the public API and the second one represents the private/internal API (for contributors).

However, I'm having some trouble getting mkdocs-python to only include the private members for the internal page. This is what I tried:

```markdown
::: mcproto.types.nbt
    options:
        filters:
            - "^_[^_]"
```

But it only works partially. This will include any top level components that start with an underscore, like functions, but it doesn't show the private components of public classes, so if I have something like:

```python
def _private_global_func():
    ...

class Foo:
    def _private_meth():
        ...
```

Only the `_private_global_func` shows here, presumably because the filter didn't match on the name of the class (`Foo`), even though that class does hold a component which would match the filter.

Other than explicitly specifying all private members, I can't figure out how to set up a filter to handle this use case. I'm not even sure if it is possible with the current implementation. If it isn't, I'd like to elevate this to a feature request.

**Comment by pawamoy** [comment, marked as answer]
Indeed, when recursively rendering members of a module, the filters will exclude your `Foo` class since it does not start with an underscore. I don't think we currently have a way to achieve what you want automatically. So my current recommendation would be to maybe add these public class names to your filters? At least while waiting for [fine-grain configuration](https://github.com/mkdocstrings/mkdocstrings/issues/658) which should unlock this interesting use-case :slightly_smiling_face:

**Reply by ItsDrike** [reply]
Hmm, adding the names of public classes does work, but there's actually another issue here too, where say I have a private class like `_NumberNBTag`, the class does show up, but its components don't, as those don't all start with an underscore. I also can't add these names into the filters directly, because some of those names are also used in the public class that has some private members, e.g.:

```python
class _PrivateClass:
    def _gets_documented(): ...  # good
    
    def doesnt_get_documented(): ...  # bad
    
class PublicClass:
    def _gets_documented(): ...  # good
    
    def doesnt_get_documneted(): ...  # good
```

Using a filter like:
```python
        filters:
            - "^_[^_]"
            - "^PublicClass$"
```

If I add a separate filter to enable `doesnt_get_documneted` here, it would get documented for the public class too, which I don't want. I guess this is simply something that needs the fine-grained configuration.

**Reply by pawamoy** [reply]
Yep, I don't see how to make this work without fine-grain configuration.

**Reply by ItsDrike** [reply]
I managed to hack my way through it via `members`, e.g.:

```python
::: mcproto.protocol.base_io
    options:
        members: [BaseAsyncWriter, BaseSyncWriter, BaseAsyncReader, BaseSyncReader]
        filters:
            - "^_[^_]"
```

**Reply by pawamoy** [reply]
Oh, nice! Yeah `members` have precedence over `filters` for the the first layer of members :slightly_smiling_face:
